### PR TITLE
Improve separation of concerns: introduce readBuildTreeRefsFromCache

### PR DESCRIPTION
**Pull request by edsko**
In `listBuildTreeRefs` we find this somewhat mysterious code:

``` haskell
listWithIgnored :: IO [BuildTreeRef]
listWithIgnored = readBuildTreeRefsFromFile $ path

listWithoutIgnored :: IO [FilePath]
listWithoutIgnored = do
  let repo = RepoLocal { repoLocalDir = takeDirectory path }
  pkgIndex <- fmap packageIndex
              . getSourcePackagesStrict verbosity $ [repo]
  return [ pkgPath | (LocalUnpackedPackage pkgPath) <-
              map packageSource . allPackages $ pkgIndex ]
```

It is not at all obvious (at least, not to me) how `listWithIgnored` and `listWithoutIgnored` do almost the same thing. In addition---and, for my current purposes, more importantly---this is the _only_ place in the entire `cabal-install` codebase where we are constructing a `Repo` out of nothing, and that's causing me a lot of grief in further refactoring I need to do to hook up the security infrastructure to the `HttpTransport` in `cabal-install`. 

This PR is very small, and changes the definition of `listWithoutIgnored` to:

``` haskell
listWithoutIgnored :: IO [FilePath]
listWithoutIgnored = fmap (map buildTreePath) $ readBuildTreeRefsFromCache verbosity path
```

which makes it a lot clearer what's going on (one reads build tree refs from the index directly, the other uses the index cache instead), and it avoids constructing a `Repo`.

While the PR might be small, reviewing it to verify that I have not in fact changed any functionality is nearly impossible without retracing my steps. So, below is the full justification for this change by deriving at the above code step by step.

# Step 1: Introduce top-level `readBuildTreeRefsFromCache`

Original code:

``` haskell
listWithoutIgnored :: IO [FilePath]
listWithoutIgnored = do
  let repo = RepoLocal { repoLocalDir = takeDirectory path }
  pkgIndex <- fmap packageIndex
              . getSourcePackagesStrict verbosity $ [repo]
  return [ pkgPath | (LocalUnpackedPackage pkgPath) <-
              map packageSource . allPackages $ pkgIndex ]
```

Moved this to a top-level function `readBuildTreeRefsFromCache`:

``` haskell
readBuildTreeRefsFromCache :: Verbosity -> FilePath -> IO [FilePath]
readBuildTreeRefsFromCache verbosity indexPath = do
    let repo = RepoLocal { repoLocalDir = takeDirectory indexPath }
    pkgIndex <- fmap packageIndex
                . getSourcePackagesStrict verbosity $ [repo]
    return [ pkgPath | (LocalUnpackedPackage pkgPath) <-
                map packageSource . allPackages $ pkgIndex ]
```

and replaced the export of `getSourcePackagesStrict` get `readBuildTreeRefsFromCache`.

# Step 2: Inline `getSourcePackages'`

Inline `getSourcePackagesStrict` (which is now obsolete)

``` haskell
readBuildTreeRefsFromCache :: Verbosity -> FilePath -> IO [FilePath]
readBuildTreeRefsFromCache verbosity indexPath = do
    let repo = RepoLocal { repoLocalDir = takeDirectory indexPath }
    pkgIndex <- fmap packageIndex $ getSourcePackages' verbosity [repo] ReadPackageIndexStrict
    return [ pkgPath | (LocalUnpackedPackage pkgPath) <-
                map packageSource . allPackages $ pkgIndex ]
```

and inline `getSourcePackages'`; `getSourcePackages'` has two cases, but the first case only applies when we pass an empty list, and since we are passing a singleton list here it doesn't apply:

``` haskell
readBuildTreeRefsFromCache :: Verbosity -> FilePath -> IO [FilePath]
readBuildTreeRefsFromCache verbosity indexPath = do
    let repo = RepoLocal { repoLocalDir = takeDirectory indexPath }
    pkgIndex <- fmap packageIndex $ get verbosity [repo] ReadPackageIndexStrict
    return [ pkgPath | (LocalUnpackedPackage pkgPath) <-
                map packageSource . allPackages $ pkgIndex ]
  where
    get verbosity repos mode = do
      info verbosity "Reading available packages..."
      pkgss <- mapM (\r -> readRepoIndex verbosity r mode) repos
      let (pkgs, prefs) = mconcat pkgss
          prefs' = Map.fromListWith intersectVersionRanges
                     [ (name, range) | Dependency name range <- prefs ]
      _ <- evaluate pkgs
      _ <- evaluate prefs'
      return SourcePackageDb {
        packageIndex       = pkgs,
        packagePreferences = prefs'
      }
```

Simplify: remove `verbosity` and `mode` argument to `get`, and replace `repos` with the singleton `repo` argument; then and get rid of `mapM` and subsequent call to `mconcat` (because it's a singleton list):

``` haskell
readBuildTreeRefsFromCache :: Verbosity -> FilePath -> IO [FilePath]
readBuildTreeRefsFromCache verbosity indexPath = do
    let repo = RepoLocal { repoLocalDir = takeDirectory indexPath }
    pkgIndex <- fmap packageIndex $ get repo
    return [ pkgPath | (LocalUnpackedPackage pkgPath) <-
                map packageSource . allPackages $ pkgIndex ]
  where
    get repo = do
      info verbosity "Reading available packages..."
      (pkgs, prefs) <- readRepoIndex verbosity repo ReadPackageIndexStrict
      let prefs' = Map.fromListWith intersectVersionRanges
                     [ (name, range) | Dependency name range <- prefs ]
      _ <- evaluate pkgs
      _ <- evaluate prefs'
      return SourcePackageDb {
        packageIndex       = pkgs,
        packagePreferences = prefs'
      }
```

Now we notice immediately after constructing the `SourcePackageDb` we call `packageIndex` to extract the first component again, so we might as well not construct it in the first place:

``` haskell
readBuildTreeRefsFromCache :: Verbosity -> FilePath -> IO [FilePath]
readBuildTreeRefsFromCache verbosity indexPath = do
    let repo = RepoLocal { repoLocalDir = takeDirectory indexPath }
    pkgIndex <- get repo
    return [ pkgPath | (LocalUnpackedPackage pkgPath) <-
                map packageSource . allPackages $ pkgIndex ]
  where
    get repo = do
      info verbosity "Reading available packages..."
      (pkgs, prefs) <- readRepoIndex verbosity repo ReadPackageIndexStrict
      evaluate pkgs
```

at which point we might as well inline `get` completely:

``` haskell
readBuildTreeRefsFromCache :: Verbosity -> FilePath -> IO [FilePath]
readBuildTreeRefsFromCache verbosity indexPath = do
    let repo = RepoLocal { repoLocalDir = takeDirectory indexPath }
    info verbosity "Reading available packages..."
    (pkgIndex, prefs) <- readRepoIndex verbosity repo ReadPackageIndexStrict
    _ <- evaluate pkgIndex
    return [ pkgPath | (LocalUnpackedPackage pkgPath) <-
                map packageSource . allPackages $ pkgIndex ]
```

# Step 3: Inline `readRepoIndex`

Inline `readRepoIndex`, immediately inlining its `verbosity` and `mode` arguments, and replacing `repo` in the body with `RepoLocal (takeDirectory indexPath)`:

``` haskell
readBuildTreeRefsFromCache :: Verbosity -> FilePath -> IO [FilePath]
readBuildTreeRefsFromCache verbosity indexPath = do
    info verbosity "Reading available packages..."
    (pkgIndex, prefs) <- readIndex
    _ <- evaluate pkgIndex
    return [ pkgPath | (LocalUnpackedPackage pkgPath) <-
                map packageSource . allPackages $ pkgIndex ]
  where
    readIndex =
      handleNotFound $ do
        warnIfIndexIsOld =<< getIndexFileAge (RepoLocal (takeDirectory indexPath))
        updateRepoIndexCache verbosity (RepoIndex (RepoLocal (takeDirectory indexPath)))
        readPackageIndexCacheFile mkAvailablePackage (RepoIndex (RepoLocal (takeDirectory indexPath))) ReadPackageIndexStrict

      where
        mkAvailablePackage pkgEntry =
          SourcePackage {
            packageInfoId      = pkgid,
            packageDescription = packageDesc pkgEntry,
            packageSource      = case pkgEntry of
              NormalPackage _ _ _ _       -> RepoTarballPackage (RepoLocal (takeDirectory indexPath)) pkgid Nothing
              BuildTreeRef  _  _ _ path _ -> LocalUnpackedPackage path,
            packageDescrOverride = case pkgEntry of
              NormalPackage _ _ pkgtxt _ -> Just pkgtxt
              _                          -> Nothing
          }
          where
            pkgid = packageId pkgEntry

        handleNotFound action = catchIO action $ \e -> if isDoesNotExistError e
          then do
            case (RepoLocal (takeDirectory indexPath)) of
              RepoRemote{..} -> warn verbosity $
                   "The package list for '" ++ remoteRepoName repoRemote
                ++ "' does not exist. Run 'cabal update' to download it."
              RepoLocal{..} -> warn verbosity $
                   "The package list for the local Repo '" ++ repoLocalDir
                ++ "' is missing. The Repo is invalid."
            return mempty
          else ioError e

        isOldThreshold = 15 --days
        warnIfIndexIsOld dt = do
          when (dt >= isOldThreshold) $ case (RepoLocal (takeDirectory indexPath)) of
            RepoRemote{..} -> warn verbosity $
                 "The package list for '" ++ remoteRepoName repoRemote
              ++ "' is " ++ shows (floor dt :: Int) " days old.\nRun "
              ++ "'cabal update' to get the latest list of available packages."
            RepoLocal{..} -> return ()
```

Now we notice a few things:

* The check for the index being too old (`warnIfIndexIsOld`) does not apply to a local repo.
* The handler for the index not being found (`handleNotFound`) is not really useful either; it is useful when the user specifies a directory which is supposed to contain a local repository, but we actually don't find an `00-index` there; however, that's not what's happening here; we are _starting_ from the location of a `00-index` and constructing a repo from it. So that check can go.
* Then we have the strange construction

    ``` haskell
    (RepoIndex (RepoLocal (takeDirectory indexPath)))
    ```

  This should of course just be

     ``` haskell
     SandboxIndex indexPath
     ```

This leaves us with

``` haskell
readBuildTreeRefsFromCache :: Verbosity -> FilePath -> IO [FilePath]
readBuildTreeRefsFromCache verbosity indexPath = do
    info verbosity "Reading available packages..."
    (pkgIndex, prefs) <- readIndex
    _ <- evaluate pkgIndex
    return [ pkgPath | (LocalUnpackedPackage pkgPath) <-
                map packageSource . allPackages $ pkgIndex ]
  where
    readIndex = do
      updateRepoIndexCache verbosity (SandboxIndex indexPath)
      readPackageIndexCacheFile mkAvailablePackage (SandboxIndex indexPath) ReadPackageIndexStrict

    mkAvailablePackage pkgEntry =
      SourcePackage {
        packageInfoId      = pkgid,
        packageDescription = packageDesc pkgEntry,
        packageSource      = case pkgEntry of
          NormalPackage _ _ _ _       -> RepoTarballPackage (RepoLocal (takeDirectory indexPath)) pkgid Nothing
          BuildTreeRef  _  _ _ path _ -> LocalUnpackedPackage path,
        packageDescrOverride = case pkgEntry of
          NormalPackage _ _ pkgtxt _ -> Just pkgtxt
          _                          -> Nothing
      }
      where
        pkgid = packageId pkgEntry
```

# Step 4: Inline `readPackageIndexCacheFile`

In `mkAvailablePackage` are constructing `SourcePackage`s with different kinds of `packageSource`s, depending on the `pkgEntry` we are given by `readPackageIndexCacheFile`. However, we then filter out only the packages that we have given `LocalUnpackagedPackage` location; in fact, we are only interested in that location itself. In other words, we are only interested in the `BuildTreeRef`s that `readPackageIndexCacheFile` returns.

Unfortunately, we cannot simplify `mkAvailablePackage` to return a `Maybe FilePath`, because although  `readPackageIndexCacheFile` is polymorphic in the type of the thing we construct from each `PackageEntry`, it does have a `Package` constraint on it. The only reason it has that construct however is that `packageIndexFromCache` builds a `PackageIndex` from what is originally a _list_ of `PackageEntry`s; but in `readBuildTreeRefsFromCache` we then immediately turn this `PackageIndex` back to a list (`allPackages`)!

So let's inline `readPackageIndexCacheFile`, immediately substituting its `mkPkg`, `index` and `mode` arguments and simplifying `myWithFile`; this gives us the following definition of `readIndex`:

``` haskell
readIndex = do
  updateRepoIndexCache verbosity (SandboxIndex indexPath)
  cache <- liftM readIndexCache $ BSS.readFile (cacheFile (SandboxIndex indexPath))
  withFile (indexFile (SandboxIndex indexPath)) ReadMode $ \indexHnd ->
    packageIndexFromCache mkAvailablePackage indexHnd cache ReadPackageIndexStrict
```

Now let's split `packageIndexFromCache` into two functions `packageListFromCache`, which does everything `packageIndexFromCache` did except turn the list into a package index, and a new `packageIndexFromCache` which is just a simple wrapper around `packageListFromCache`:

``` haskell
packageIndexFromCache :: Package pkg
                      => (PackageEntry -> pkg)
                      -> Handle
                      -> Cache
                      -> ReadPackageIndexMode
                      -> IO (PackageIndex pkg, [Dependency])
packageIndexFromCache mkPkg hnd cache mode = do
     (pkgs, prefs) <- packageListFromCache mkPkg hnd cache mode
     pkgIndex <- evaluate $ PackageIndex.fromList pkgs
     return (pkgIndex, prefs)

-- | Read package list
--
-- The result packages (though not the preferences) are guaranteed to be listed
-- in the same order as they are in the tar file (because later entries in a tar
-- file mask earlier ones).
packageListFromCache :: (PackageEntry -> pkg)
                     -> Handle
                     -> Cache
                     -> ReadPackageIndexMode
                     -> IO ([pkg], [Dependency])
packageListFromCache mkPkg hnd Cache{..} mode = accum mempty [] cacheEntries
  where
    accum srcpkgs prefs [] = return (reverse srcpkgs, prefs)

    -- other cases for 'accum' as before
```    

We can now use `packageListFromCache` in `readIndex`, leaving us with

``` haskell
readBuildTreeRefsFromCache :: Verbosity -> FilePath -> IO [FilePath]
readBuildTreeRefsFromCache verbosity indexPath = do
    info verbosity "Reading available packages..."
    (pkgIndex, prefs) <- readIndex
    _ <- evaluate pkgIndex
    return [ pkgPath | (LocalUnpackedPackage pkgPath) <-
                map packageSource $ pkgIndex ]
  where
    readIndex = do
      updateRepoIndexCache verbosity (SandboxIndex indexPath)
      cache <- liftM readIndexCache $ BSS.readFile (cacheFile (SandboxIndex indexPath))
      withFile (indexFile (SandboxIndex indexPath)) ReadMode $ \indexHnd ->
        packageListFromCache mkAvailablePackage indexHnd cache ReadPackageIndexStrict

    mkAvailablePackage pkgEntry =
      SourcePackage {
        packageInfoId      = pkgid,
        packageDescription = packageDesc pkgEntry,
        packageSource      = case pkgEntry of
          NormalPackage _ _ _ _       -> RepoTarballPackage (RepoLocal (takeDirectory indexPath)) pkgid Nothing
          BuildTreeRef  _  _ _ path _ -> LocalUnpackedPackage path,
        packageDescrOverride = case pkgEntry of
          NormalPackage _ _ pkgtxt _ -> Just pkgtxt
          _                          -> Nothing
      }
      where
        pkgid = packageId pkgEntry
```

# Step 5: Simplify `mkAvailablePackage`

We can now simplify `mkAvailablePackage` to return the build tree references only, which is what we're really interested in; renaming `mkAvailablePackage` to `buildTreeRef` we get:

``` haskell
readBuildTreeRefsFromCache :: Verbosity -> FilePath -> IO [FilePath]
readBuildTreeRefsFromCache verbosity indexPath = do
    info verbosity "Reading available packages..."
    (pkgIndex, prefs) <- readIndex
    _ <- evaluate pkgIndex
    return $ catMaybes pkgIndex
  where
    readIndex = do
      updateRepoIndexCache verbosity (SandboxIndex indexPath)
      cache <- liftM readIndexCache $ BSS.readFile (cacheFile (SandboxIndex indexPath))
      withFile (indexFile (SandboxIndex indexPath)) ReadMode $ \indexHnd ->
        packageListFromCache buildTreeRef indexHnd cache ReadPackageIndexStrict

    buildTreeRef pkgEntry =
      case pkgEntry of
         NormalPackage _ _ _ _ -> Nothing
         BuildTreeRef _ _ _ path _ -> Just path
```

# Step 6: Final cleanup

We can inline `readIndex`, and we get

``` haskell
readBuildTreeRefsFromCache :: Verbosity -> FilePath -> IO [FilePath]
readBuildTreeRefsFromCache verbosity indexPath = do
    info verbosity "Reading available packages..."
    updateRepoIndexCache verbosity (SandboxIndex indexPath)
    cache <- liftM readIndexCache $ BSS.readFile (cacheFile (SandboxIndex indexPath))
    (pkgIndex, _prefs) <- withFile (indexFile (SandboxIndex indexPath)) ReadMode $ \indexHnd ->
      packageListFromCache buildTreeRef indexHnd cache ReadPackageIndexStrict
    return $ catMaybes pkgIndex
  where
    buildTreeRef pkgEntry =
      case pkgEntry of
         NormalPackage _ _ _ _ -> Nothing
         BuildTreeRef _ _ _ path _ -> Just path
```

Finally, we can recognize that most of this is completely generic code for reading an index cache, so we can split it out:

``` haskell
readCacheStrict :: Verbosity -> Index -> (PackageEntry -> pkg) -> IO ([pkg], [Dependency])
readCacheStrict verbosity index mkPkg = do
    updateRepoIndexCache verbosity index
    cache <- liftM readIndexCache $ BSS.readFile (cacheFile index)
    withFile (indexFile index) ReadMode $ \indexHnd ->
      packageListFromCache mkPkg indexHnd cache ReadPackageIndexStrict

-- | Read build tree references from an index cache
readBuildTreeRefsFromCache :: Verbosity -> FilePath -> IO [FilePath]
readBuildTreeRefsFromCache verbosity indexPath = do
    (mRefs, _prefs) <- readCacheStrict verbosity (SandboxIndex indexPath) buildTreeRef
    return (catMaybes mRefs)
  where
    buildTreeRef pkgEntry =
      case pkgEntry of
         IndexUtils.NormalPackage _ _ _ _ -> Nothing
         IndexUtils.BuildTreeRef _ _ _ path _ -> Just path
```

At this point the difference between `listWithIgnored` and `listWithoutIgnored` has become a _lot_ clearer:

``` haskell
listWithIgnored :: IO [BuildTreeRef]
listWithIgnored = readBuildTreeRefsFromFile $ path

listWithoutIgnored :: IO [FilePath]
listWithoutIgnored = readBuildTreeRefsFromCache verbosity path
```

The only weird thing is that that `readBuildTreeRefsFromCache` now returns a list of `FilePath` rather than `BuildTreeRef`; but we _do_ have these `BuildTreeRef`s; we threw them away in `readBuildTreeRefsFromCache`. I have no idea why we do that, but that's what the original code in `listBuildTreeRefs` did. It would be nicer to make that a little bit more explicit though, so let's change it:

``` haskell
-- | Read build tree references from an index cache
readBuildTreeRefsFromCache :: Verbosity -> FilePath -> IO [BuildTreeRef]
readBuildTreeRefsFromCache verbosity indexPath = do
    (mRefs, _prefs) <- readCacheStrict verbosity (SandboxIndex indexPath) buildTreeRef
    return (catMaybes mRefs)
  where
    buildTreeRef pkgEntry =
      case pkgEntry of
         IndexUtils.NormalPackage _ _ _ _ -> Nothing
         IndexUtils.BuildTreeRef typ _ _ path _ -> Just $ BuildTreeRef typ path
```

which gives us this final definitions in `listBuildTreeRefs`:

``` haskell
listWithIgnored :: IO [BuildTreeRef]
listWithIgnored = readBuildTreeRefsFromFile path

listWithoutIgnored :: IO [FilePath]
listWithoutIgnored = fmap (map buildTreePath)
                   $ readBuildTreeRefsFromCache verbosity path

### CODE REVIEW

**Comment by ezyang** [comment]
Hello @edsko, very nice analysis. Perhaps you could top it off by adding some comments saying, intuitively, what these variables are supposed to represent?

**Comment by edsko** [comment]
@ezyang To be honest, I have no idea :) I think the sandbox stuff (ab)uses the tarfile to store references (symlinks) to local packages, which it calls "build references" -- but how this interacts with the cache, exactly, I don't know. The analysis above to me was just a means to an end :) But perhaps @dcoutts knows.

**Comment by 23Skidoo** [comment]
LGTM, thanks for explaining what you did in such detail.

@ezyang 

This is explained in the comment for `ListIgnoredBuildTreeRefs`:

> ```
-- | A build tree ref can become ignored if the user later adds a build tree ref
-- with the same package ID. We display ignored build tree refs when the user
-- runs 'cabal sandbox list-sources', but do not look at their timestamps in
-- 'reinstallAddSourceDeps'.
data ListIgnoredBuildTreeRefs = ListIgnored | DontListIgnored
```

@edsko 

> I think the sandbox stuff (ab)uses the tarfile to store references (symlinks) to local packages, which it calls "build references"

Basically yes, this is how `add-source` is implemented. When the user does something like this:

```
$ cabal sandbox add-source /path/to/foo-1.0
$ cabal sandbox add-source /some/other/path/to/foo-1.0
```

a build tree ref becomes ignored, and is no longer present in the cache, though still present in the index itself.
